servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java

```diff
@@ -258,19 +295,31 @@ protected Single<StreamingHttpResponse> request(final StreamingHttpRequester del
             if (responseMapper != null) {
                 single = single.flatMap(resp -> {
                     final HttpResponseException exception = responseMapper.apply(resp);
-                    return (exception != null ?
-                            // Drain response payload body before discarding it:
-                            resp.payloadBody().ignoreElements().onErrorComplete()
-                                    .concat(Single.<StreamingHttpResponse>failed(exception)) :
-                            Single.succeeded(resp))
-                            .shareContextOnSubscribe();
+                    if (exception == null) {
+                        return Single.succeeded(resp).shareContextOnSubscribe();
+                    }
+                    if (returnFailedResponses) {
+                        return Single.failed(new WrappedResponseException(resp, exception));
+                    } else {
+                        return drainResponse(resp).concat(Single.<StreamingHttpResponse>failed(exception));
+                    }
                 });
             }
 
             // 1. Metadata is shared across retries
             // 2. Publisher state is restored to original state for each retry
             // duplicatedRequest isn't used below because retryWhen must be applied outside the defer operator for (2).
-            return single.retryWhen(retryStrategy(request, executionContext(), true));
+            single = single.retryWhen(retryStrategy(request, executionContext(), true));
+            if (returnFailedResponses) {
+                single = single.onErrorResume(t -> {
+                    if (t instanceof WrappedResponseException) {
```

Comment:
Consider using `onErrorResume` overload that takes exception class as the first arg